src/elf/gnu_hash.rs

```diff
@@ -1,143 +1,206 @@
+/// used to compute a bucket index.
+///
+/// Consequently, if the hashing function returns the value `x` for some name,
+/// `buckets[x % nbuckets]` gives an index, `y`, into both the symbol table
+/// and the chain table.
 pub fn hash(symbol: &str) -> u32 {
     const HASH_SEED: u32 = 5381;
-    let mut hash = HASH_SEED;
-    for b in symbol.as_bytes() {
-        hash = hash
-            .wrapping_mul(33)
-            .wrapping_add(u32::from(*b));
-    }
-    hash
+    symbol.bytes().fold(HASH_SEED, |hash, b| {
+        hash.wrapping_mul(33).wrapping_add(u32::from(b))
+    })
 }
 
 #[cfg(test)]
 mod tests {
     use super::hash;
     #[test]
     fn test_hash() {
-        assert_eq!(hash("")             , 0x00001505);
-        assert_eq!(hash("printf")       , 0x156b2bb8);
-        assert_eq!(hash("exit")         , 0x7c967e3f);
-        assert_eq!(hash("syscall")      , 0xbac212a0);
+        assert_eq!(hash(""), 0x00001505);
+        assert_eq!(hash("printf"), 0x156b2bb8);
+        assert_eq!(hash("exit"), 0x7c967e3f);
+        assert_eq!(hash("syscall"), 0xbac212a0);
         assert_eq!(hash("flapenguin.me"), 0x8ae9f18e);
     }
 }
 
 macro_rules! elf_gnu_hash_impl {
     ($size:ty) => {
-
-        use core::slice;
-        use core::mem;
-        use crate::strtab::Strtab;
         use super::sym;
+        use crate::strtab::Strtab;
+        use core::mem;
+        use core::slice;
+
+        const BYTE_SIZE: u32 = 8;
+        /// Size of a bits mask in bloom filter
+        const ELFCLASS_BITS: u32 = mem::size_of::<$size>() as u32 * BYTE_SIZE;
 
-        pub struct GnuHash<'process> {
+        pub struct GnuHash<'a> {
+            /// Number of GNU hash `buckets`
             nbuckets: u32,
-            symindex: usize,
+            /// Index of the first symbol in the .dynsym table which is accessible with
+            /// the hash table
+            symindex: u32,
+            /// Number of elements in the bloom filter, minus 1
+            maskwords: u32,
+            /// Shift count used in the bloom filter
             shift2: u32,
-            maskbits: u32,
-            bloomwords: &'process [$size], // either 32 or 64 bit masks, depending on platform
-            maskwords_bitmask: u32,
-            buckets: &'process [u32],
-            hashvalues: &'process [u32],
-            symtab: &'process [sym::Sym],
+            /// 2 bit bloom filter on `chains`
+            // Either 32 or 64-bit depending on the class of object
+            bloom_filter: &'a [$size], // => bloom_filter[maskwords]
+            /// GNU hash table bucket array; indexes start at 0. This array holds symbol
+            /// table indexes and contains the index of hashes in `chains`
+            buckets: &'a [u32], // => buckets[nbuckets]
+            /// Hash values; indexes start at 0. This array holds symbol table indexes.
+            chains: &'a [u32], // => chains[total_dynsyms - symindex]
+        }
+
+        impl<'a> core::fmt::Debug for GnuHash<'a> {
```

Comment:
thank you for this!

Reply:
I added it in process of debugging this PR.